File: packages/catsproto/lib/src/connection.dart
import 'dart:async';
import 'dart:convert';
import 'dart:io';
import 'dart:math';

import 'endpoint.dart';
import 'generated/codec.g.dart';
import 'generated/commands.g.dart';
import 'generated/wire.g.dart';

/// The socket a [CatsConnection] drives.
///
/// Abstracted so a test can replay a recorded transcript through the exact same
/// fold the live client uses. Every widget test, every golden, and demo mode
/// run on the same code path as the real thing — the alternative is a mock that
/// agrees with the client and disagrees with the server.
abstract interface class CatsSocket {
  Stream<String> get messages;

  void send(String text);

  Future<void> close();
}

/// A real WebSocket, with the two hooks `package:web_socket_channel` cannot
/// give us: an `Authorization` header on the upgrade, and a certificate
/// callback. Both are load-bearing — the bearer token IS the credential, and
/// catway serves a self-signed certificate that only a pin can validate.
class WebSocketCatsSocket implements CatsSocket {
  WebSocketCatsSocket._(this._ws);

  static Future<WebSocketCatsSocket> connect({
    required Endpoint endpoint,
    required String token,
    String? storedPin,
    required void Function(CertDecision) onCertDecision,
    Duration pingInterval = const Duration(seconds: 20),
  }) async {
    final ws = await WebSocket.connect(
      endpoint.wsUri.toString(),
      // No Origin header is sent. gwauth.OriginOK returns true for an empty
      // Origin ("non-browser client… auth is still enforced"), so a native
      // client passes the check for free — the header exists to stop a hostile
      // PAGE, and there is no page here.
      headers: {'Authorization': 'Bearer $token'},
      customClient: endpoint.tls
          ? pinnedHttpClient(
              endpoint: endpoint,
              storedPin: storedPin,
              onDecision: onCertDecision,
            )
          : null,
    );
    // Dart replies to server pings in the platform layer; this is the other
    // direction, so a NAT that drops an idle flow gets traffic to keep. catway
    // pings at 30s and reads with a 90s deadline, so 20s here is comfortably
    // inside its window without being chatty on cellular.
    ws.pingInterval = pingInterval;
    return WebSocketCatsSocket._(ws);
  }

  final WebSocket _ws;

  @override
  Stream<String> get messages => _ws.map((e) => e is String ? e : '');

  @override
  void send(String text) => _ws.add(text);

  @override
  Future<void> close() => _ws.close();
}

/// Thrown when a command comes back `ok: false`.
class CatsCommandError implements Exception {
  CatsCommandError(this.command, this.message);

  final String command;
  final String message;

  @override
  String toString() => 'CatsCommandError($command): $message';
}

/// Thrown when a command is still outstanding at disconnect. Distinct from a
/// timeout so the UI can say "reconnecting" instead of "the server is slow".
class CatsDisconnectedError implements Exception {
  CatsDisconnectedError(this.command);

  final String command;

  @override
  String toString() =>
      'CatsDisconnectedError($command): the socket closed first';
}

/// Thrown by [CatsConnection.send] for a message a viewer must never emit.
class ViewerModeViolation implements Exception {
  ViewerModeViolation(this.what);

  final String what;

  @override
  String toString() =>
      'ViewerModeViolation: a cats client never sends $what — it would reshape '
      'the desktop it is looking at';
}

/// One live session with one catway.
///
/// # The phone never resizes the desktop
///
/// catway's `registerConn` takes the session grid from the FIRST `init` that
/// declares one, and every connection shares that grid. A phone honestly
/// reporting 40×20 reflows every pane for everybody. Four independent layers
/// stop that, and this class is two of them:
///
///   1. [_handshake] builds the `Init` itself, with hardcoded zeros and
///      `viewer: true`. App code has no way to supply one — except the
///      workspace it wants to look through, which says nothing about size.
///   2. [send] refuses `Resize` outright.
///   3. the generator marks `Resize` `@Deprecated`, and analysis_options.yaml
///      promotes `deprecated_member_use` to an error.
///   4. a source test fails on any `Resize(` or non-zero `cols:` under lib/.
///
/// Belt and braces is warranted here because it is the one failure mode where a
/// bug in the phone breaks the user's desktop.
///
/// [followWorkspace] adds a fifth of the same kind for the same reason: it
/// refuses to send `workspace.focus` unless the server advertises
/// [Caps.window]. With that capability the command moves only this connection's
/// view; without it, it is the old session-wide switch and would drag every
/// window at the desk along.
class CatsConnection with CatsCommands implements CatsCommandTransport {
  CatsConnection({
    required this.endpoint,
    required CatsSocket socket,
    this.defaultTimeout = const Duration(seconds: 30),
    this.maxTimeout = const Duration(minutes: 10),
    String workspace = '',
  }) : _socket = socket,
       _pinnedWorkspace = workspace {
    _subscription = _socket.messages.listen(
      _onText,
      onError: (Object e) => _fail(e),
      onDone: () => _fail(CatsDisconnectedError('<socket closed>')),
      cancelOnError: false,
    );
    // A caller that never awaits [welcome] — the common case once the session
    // is up — must not turn a disconnect into an unhandled async error that
    // takes down the zone. ignore() registers a discarding listener; a real
    // await still receives the error, because a Future can have several.
    _welcome.future.ignore();
    _handshake();
  }

  final Endpoint endpoint;

  /// Bounds an ordinary command. `pane.wait_for_output` overrides it and can
  /// legitimately run to [maxTimeout] — app.MaxWaitTimeout is ten minutes, and
  /// a client deadline shorter than that would abandon a wait the server is
  /// still faithfully serving.
  final Duration defaultTimeout;
  final Duration maxTimeout;

  final CatsSocket _socket;
  late final StreamSubscription<String> _subscription;

  final _incoming = StreamController<Object>.broadcast();
  final Map<String, _Pending> _pending = {};
  final _welcome = Completer<Welcome>();
  int _nextId = 0;
  bool _closed = false;

  /// Every decoded down-message, in arrival order. Unknown types never appear —
  /// [decodeDown] drops them, which is the protocol's rule.
  Stream<Object> get messages => _incoming.stream;

  /// Completes with the server's `welcome`, or with an error when the server
  /// rejected the handshake (a protocol-version mismatch) or the socket died
  /// first.
  Future<Welcome> get welcome => _welcome.future;

  /// The capabilities the server advertised, once [welcome] has arrived. Empty
  /// before that — and an empty set is a real answer, not "unknown": a server
  /// old enough to send no caps honours none of them.
  Set<String> get caps => _caps;
  Set<String> _caps = const {};

  bool get isClosed => _closed;

  void _handshake() {
    // The zeros are the point. catway's registerConn skips the o.area
    // assignment when cols/rows are 0, and skips the cell metrics when those
    // are 0 — two separate guards, both of which a viewer must clear. Sending
    // `viewer: true` as well is belt and braces against a server that honours
    // one and forgets the other.
    //
    // `workspace` is the one field the app gets to fill in, and it is not a
    // hole in the rule above: a workspace is which window this connection looks
    // THROUGH, not a claim about its size. Empty — the default — is "follow the
    // primary view", which is what a phone that has not picked a window wants.
    // It rides the handshake rather than a command afterwards so a reconnect
    // comes back on the same window instead of flicking to the desk's for a
    // frame; see [followWorkspace] for the live half.
    final init = Init(
      v: kProtocolVersion,
      cols: 0,
      rows: 0,
      dpr: 0,
      cellWPx: 0,
      cellHPx: 0,
      viewer: true,
      workspace: _pinnedWorkspace,
    );
    _socket.send(jsonEncode(init.toJson()));
  }

  // --- which window this connection looks through -----------------------------
  //
  // A connection is a view. A desktop window pins its workspace with `?ws=`;
  // this client pins it in the handshake and can move it afterwards. The state
  // is here rather than in CatsSession because it is a fact about the SOCKET —
  // it dies with it, and the app hands it to the next CatsConnection to have a
  // reconnect land on the same window.
  //
  // What the server is showing is a different question, and the session answers
  // that one (CatsSession.viewWorkspace) from the layout it actually received.
  // A pin can be stale — the workspace it named can be closed from the desk —
  // and the server falls back silently when it is, so a UI that reports the pin
  // as "what you are watching" would be lying at exactly the wrong moment.

  /// The workspace this connection asked to be pinned to, or '' when it is
  /// following the primary view — whichever desktop window was touched last.
  String get pinnedWorkspace => _pinnedWorkspace;
  String _pinnedWorkspace;

  /// Whether this connection is following the primary view rather than holding
  /// one window.
  bool get followsPrimaryView => _pinnedWorkspace.isEmpty;

  /// Pins this connection to one desktop window's workspace.
  ///
  /// This is a viewer-safe command and the fifth layer of the class doc's
  /// guard is what makes it one. On a server advertising [Caps.window],
  /// `workspace.focus` moves ONLY the connection that sent it — nothing at the
  /// desk changes, which is why picking a window to watch does not need the
  /// "explicit confirmed gesture" that `agent.focus` or `pane.zoom` does. On a
  /// server WITHOUT that capability the same command is a session mutation and
  /// would switch every window at the desk, so it is refused here rather than
  /// sent and hoped about — the one failure mode where a bug in the phone
  /// rearranges somebody's desktop.
  ///
  /// Awaits [welcome] first: the capability set is not known before it, and an
  /// empty set is a real answer ("honours none") that must not be read as a
  /// refusal on a server that simply has not replied yet.
  Future<void> followWorkspace(String workspaceId) async {
    await _requireWindowCap('workspace.focus');
    await workspaceFocus(WorkspaceParams(id: workspaceId));
    // Only after the server said ok: a pin the server rejected (an id closed
    // between the census and the tap) must not survive into the next
    // handshake, where it would be silently fallen back a second time.
    _pinnedWorkspace = workspaceId;
  }

  /// Releases the pin: this connection follows the primary view again.
  ///
  /// The return leg of [followWorkspace] — an empty id is "follow the primary",
  /// the state [Init.workspace] leaves a connection in by being absent. A
  /// server too old to know it answers `ok: false` and the pin stands, which is
  /// a detectable no rather than a silent one.
  Future<void> followPrimaryView() async {
    await _requireWindowCap('workspace.focus');
    await workspaceFocus(const WorkspaceParams(id: ''));
    _pinnedWorkspace = '';
  }

  Future<void> _requireWindowCap(String what) async {
    await welcome;
    if (!_caps.contains(Caps.window)) {
      // Phrased to complete ViewerModeViolation's own sentence, which already
      // ends "— it would reshape the desktop it is looking at".
      throw ViewerModeViolation(
        '$what to a server that does not advertise "${Caps.window}", where it '
        'is a session-wide switch',
      );
    }
  }

  /// Sends one up-message.
  ///
  /// Refuses `Resize` regardless of what the caller intended: see the class doc.
  void send(Object message) {
    if (_closed) throw CatsDisconnectedError('<send>');
    // The type test is the ONLY use of Resize in this package, and it exists to
    // refuse it. Naming a deprecated type in order to reject it is exactly the
    // case the lint cannot distinguish, so it is silenced here and nowhere else.
    // ignore: deprecated_member_use_from_same_package
    if (message is Resize) throw ViewerModeViolation('resize');
    if (message is Init) {
      throw ViewerModeViolation(
        'a second init (the handshake is ours to build)',
      );
    }
    final json = switch (message) {
      final Key m => m.toJson(),
      final Mouse m => m.toJson(),
      final Paste m => m.toJson(),
      final Image m => m.toJson(),
      final Cmd m => m.toJson(),
      _ => throw ArgumentError('not an up-message: ${message.runtimeType}'),
    };
    _socket.send(jsonEncode(json));
  }

  @override
  Future<Object?> invoke(String name, Object? params) {
    if (_closed) return Future.error(CatsDisconnectedError(name));
    final id = 'c${_nextId++}';
    final pending = _Pending(name, Completer<Object?>());
    _pending[id] = pending;

    _socket.send(jsonEncode(Cmd(id: id, name: name, params: params).toJson()));

    // Every call is bounded. A wait that resolves only on a server event is the
    // one case where a long deadline is correct rather than lazy — but even
    // that one has a deadline, because a leaked completer is a spinner that
    // never stops.
    final limit = name == CmdName.paneWaitForOutput
        ? maxTimeout
        : defaultTimeout;
    pending.timer = Timer(limit, () {
      if (_pending.remove(id) != null) {
        pending.completer.completeError(
          TimeoutException('$name did not answer within $limit'),
        );
      }
    });
    return pending.completer.future;
  }

  void _onText(String text) {
    final json = decodeFrame(text);
    if (json == null) return; // malformed: drop the message, keep the socket

    final msg = decodeDown(json);
    if (msg == null) return; // unknown "t": the protocol says ignore it

    if (msg is Welcome) {
      _caps = Set.unmodifiable(msg.caps);
      if (!_welcome.isCompleted) {
        if (msg.error.isNotEmpty) {
          _welcome.completeError(CatsCommandError('init', msg.error));
        } else {
          _welcome.complete(msg);
        }
      }
    }
    if (msg is CmdResult) {
      final pending = _pending.remove(msg.id);
      if (pending != null) {
        pending.timer?.cancel();
        if (msg.ok) {
          pending.completer.complete(msg.data);
        } else {
          pending.completer.completeError(
            CatsCommandError(pending.name, msg.error),
          );
        }
      }
      // A CmdResult with no pending entry is not an error: it is the reply to a
      // call that already timed out. Falling through publishes it on [messages]
      // anyway, where a debug view can see it.
    }
    if (!_incoming.isClosed) _incoming.add(msg);
  }

  /// Fails every outstanding call. Called on socket error and on close.
  ///
  /// Every pending completer must fail, without exception. The alternative is
  /// a Future nobody ever completes, which in Flutter is a spinner that spins
  /// until the app is killed.
  void _fail(Object error) {
    if (_closed) return;
    _closed = true;
    for (final entry in _pending.entries) {
      entry.value.timer?.cancel();
      if (!entry.value.completer.isCompleted) {
        entry.value.completer.completeError(
          error is CatsDisconnectedError
              ? CatsDisconnectedError(entry.value.name)
              : error,
        );
      }
    }
    _pending.clear();
    if (!_welcome.isCompleted) _welcome.completeError(error);
    if (!_incoming.isClosed) _incoming.close();
  }

  Future<void> close() async {
    _fail(CatsDisconnectedError('<close>'));
    await _subscription.cancel();
    await _socket.close();
  }
}

class _Pending {
  _Pending(this.name, this.completer);

  final String name;
  final Completer<Object?> completer;
  Timer? timer;
}

/// Reconnect backoff: 500 ms doubling to 30 s, with ±20% jitter.
///
/// The jitter is not decoration. Every phone on a network that dropped comes
/// back at the same moment, and an unjittered ladder has them all knocking in
/// lockstep — which is how a home server that just restarted gets a thundering
/// herd instead of a queue.
class Backoff {
  Backoff({
    this.initial = const Duration(milliseconds: 500),
    this.ceiling = const Duration(seconds: 30),
    Random? random,
  }) : _random = random ?? Random();

  final Duration initial;
  final Duration ceiling;
  final Random _random;

  int _attempt = 0;

  /// Call on a successful connect, on resume, on a network change, and on
  /// pull-to-refresh. A user who pulled to refresh is telling you they think it
  /// should work now; making them wait out a 30 s ladder is answering an
  /// explicit request with a shrug.
  void reset() => _attempt = 0;

  Duration next() {
    final scaled = initial.inMilliseconds * (1 << _attempt.clamp(0, 20));
    final capped = min(scaled, ceiling.inMilliseconds);
    _attempt++;
    final jitter = 1.0 + (_random.nextDouble() * 0.4 - 0.2);
    return Duration(milliseconds: (capped * jitter).round());
  }
}

File: app/connection.go
package catsapp

import (
	"context"
	"errors"
	"strings"
	"sync"
	"time"

	"github.com/rohanthewiz/cats-mobile/internal/catsclient"
	"github.com/rohanthewiz/cats-mobile/internal/store"
	"github.com/rohanthewiz/cats/wire"
)

// Connection is the app's one live link to a catway: the reconnect loop,
// the Session it folds messages into, and the status the screens show.
//
// # The shape, as the plan drew it
//
//	WS reader goroutine ──► c.mu.Lock(); session.Apply(msg); Unlock()
//	                     ──► notify()   (ctx.RequestRender: the render manager
//	                                     coalesces a burst of PaneDiffs into
//	                                     one pass, because a pass drains the
//	                                     dirty flag rather than a queue)
//
// The Session has no lock of its own (see catsclient's package doc); this
// type holds it behind mu and takes that lock around Apply and around every
// read a render pass makes through Read. The connection lives at the root
// scope for the life of the process, not in a navigation frame: popping the
// pane screen must stop nothing.
//
// # Generations
//
// Connect may be called again (a new pairing, a switch of endpoint, a retry)
// while an old loop is still winding down. Every loop carries a generation
// number, and a message or status write from a loop that is no longer the
// current one is dropped. Without this an old socket's final "disconnected"
// could overwrite the new socket's "connected".
type Connection struct {
	mu       sync.Mutex
	session  *catsclient.Session
	conn     *catsclient.Conn
	status   Status
	lastErr  error
	endpoint catsclient.Endpoint
	// certSeen is the fingerprint the server presented when the pin refused
	// it, for the mismatch screen to show.
	certSeen string
	// agentsAt is when the last agents rollup arrived. since_ms is an age as
	// of the rollup, and the rollup only goes out on a change, so the label
	// has to tick from here (see AgentAge).
	agentsAt time.Time
	gen      int
	cancel   context.CancelFunc
	// wake is signalled by Retry so a loop sitting in its backoff wait tries
	// again now. Buffer of one: a second Retry during the same wait is the
	// same request.
	wake chan struct{}

	notify func()
	dialer Dialer
	store  *store.Store
}

// Status is where the connection is, for the banner and the More screen.
type Status int

const (
	// StatusIdle: no endpoint, nothing dialled.
	StatusIdle Status = iota
	// StatusConnecting: dialling, or waiting for the welcome.
	StatusConnecting
	StatusConnected
	// StatusReconnecting: the socket dropped and the backoff ladder is
	// running. The last session state stays on screen, marked stale.
	StatusReconnecting
	// StatusCertMismatch: a pin exists and the server's certificate is not
	// it. A hard stop, never retried: the only way out is "forget device"
	// and a fresh pairing.
	StatusCertMismatch
	// StatusUnauthorized: the server refused the credential. The session
	// expired or was revoked; re-pair.
	StatusUnauthorized
	// StatusVersionMismatch: the server speaks a protocol version this app
	// does not. "App update required."
	StatusVersionMismatch
)

func (s Status) String() string {
	switch s {
	case StatusIdle:
		return "Not connected"
	case StatusConnecting:
		return "Connecting…"
	case StatusConnected:
		return "Connected"
	case StatusReconnecting:
		return "Reconnecting…"
	case StatusCertMismatch:
		return "Certificate changed"
	case StatusUnauthorized:
		return "Session expired"
	case StatusVersionMismatch:
		return "App update required"
	}
	return "Unknown"
}

// Dialer opens the transport. It is the seam the tests use to replace the
// network with a fake socket, so every screen test runs the real fold over
// a recorded transcript rather than a mock of the session.
type Dialer func(ctx context.Context, endpoint catsclient.Endpoint, opts catsclient.DialOptions) (catsclient.Socket, error)

func newConnection(st *store.Store, dialer Dialer) *Connection {
	return &Connection{
		session: catsclient.NewSession(),
		store:   st,
		dialer:  dialer,
		wake:    make(chan struct{}, 1),
		notify:  func() {},
	}
}

// SetNotify installs the render trigger. Called once from Bind.
func (c *Connection) SetNotify(fn func()) {
	c.mu.Lock()
	defer c.mu.Unlock()
	if fn != nil {
		c.notify = fn
	}
}

// Connect starts (or restarts) the reconnect loop against endpoint. Any
// previous loop is cancelled; its socket closes on its own goroutine, so this
// never blocks a tap handler.
func (c *Connection) Connect(endpoint catsclient.Endpoint) {
	c.mu.Lock()
	if c.cancel != nil {
		c.cancel()
	}
	ctx, cancel := context.WithCancel(context.Background())
	c.cancel = cancel
	c.gen++
	gen := c.gen
	c.endpoint = endpoint
	c.status = StatusConnecting
	c.lastErr = nil
	c.certSeen = ""
	c.conn = nil
	c.session = catsclient.NewSession()
	c.mu.Unlock()
	c.notify()

	go c.run(ctx, gen, endpoint)
}

// Disconnect stops the loop and drops the session. Used by "forget device"
// and by switching endpoints.
func (c *Connection) Disconnect() {
	c.mu.Lock()
	if c.cancel != nil {
		c.cancel()
		c.cancel = nil
	}
	c.gen++
	c.status = StatusIdle
	c.conn = nil
	c.session = catsclient.NewSession()
	c.mu.Unlock()
	c.notify()
}

// Retry asks a loop that is waiting out its backoff to try now, and restarts
// a loop that stopped on a hard error (the user tapped "try again" after
// fixing something). A user who asks is saying they think it should work;
// making them wait out a 30 s ladder answers that with a shrug.
func (c *Connection) Retry() {
	c.mu.Lock()
	status := c.status
	endpoint := c.endpoint
	running := c.cancel != nil
	c.mu.Unlock()

	if running && status == StatusReconnecting {
		select {
		case c.wake <- struct{}{}:
		default:
		}
		return
	}
	if endpoint.Host != "" {
		c.Connect(endpoint)
	}
}

// run is one generation of the reconnect loop. It returns when ctx is
// cancelled or on a hard stop (certificate mismatch, refused credential,
// protocol mismatch); every other failure walks the backoff ladder.
func (c *Connection) run(ctx context.Context, gen int, endpoint catsclient.Endpoint) {
	backoff := catsclient.NewBackoff(nil)
	for {
		if !c.setStatus(gen, StatusConnecting, nil) {
			return
		}

		token, _, _ := c.store.ReadToken(endpoint.ID)
		pin, _, _ := c.store.ReadPin(endpoint.ID)
		if pin == "" {
			// The pairing QR carried one; the dial should check against it
			// even before the first successful connect has stored it.
			pin = endpoint.PinnedSHA256
		}

		// The decision callback runs synchronously inside the TLS handshake,
		// so by the time the dial returns it has either fired or there was
		// no self-signed certificate to decide about.
		var decision *catsclient.CertDecision
		socket, err := c.dialer(ctx, endpoint, catsclient.DialOptions{
			Token:     token,
			StoredPin: pin,
			OnCertDecision: func(d catsclient.CertDecision) {
				decision = &d
			},
		})
		if ctx.Err() != nil {
			if socket != nil {
				_ = socket.Close()
			}
			return
		}
		if decision != nil && decision.Verdict == catsclient.VerdictFirstUse {
			// Trust on first use: remember what we saw, so a later change is
			// a mismatch rather than another first use.
			_ = c.store.WritePin(endpoint.ID, decision.Fingerprint)
		}
		if err != nil {
			switch {
			case errors.Is(err, catsclient.ErrCertMismatch):
				seen := ""
				if decision != nil {
					seen = decision.Fingerprint
				}
				c.mu.Lock()
				if gen == c.gen {
					c.certSeen = seen
				}
				c.mu.Unlock()
				c.setStatus(gen, StatusCertMismatch, err)
				return
			case isUnauthorized(err):
				c.setStatus(gen, StatusUnauthorized, err)
				return
			}
			if !c.setStatus(gen, StatusReconnecting, err) {
				return
			}
			if !c.wait(ctx, backoff.Next()) {
				return
			}
			continue
		}

		// A new socket: the grids from the old one are corruption waiting to
		// happen (diff indices relative to another frame's width), so the
		// session drops them before the first message can arrive.
		c.mu.Lock()
		if gen != c.gen {
			c.mu.Unlock()
			_ = socket.Close()
			return
		}
		c.session.ResetForNewSocket()
		conn := catsclient.New(socket, catsclient.Options{
			Endpoint:  endpoint,
			OnMessage: func(msg any) { c.onMessage(gen, msg) },
		})
		c.conn = conn
		c.mu.Unlock()

		if _, err := conn.Welcome(ctx); err != nil {
			_ = conn.Close()
			if ctx.Err() != nil {
				return
			}
			var mismatch *catsclient.VersionMismatchError
			var refused *catsclient.CommandError
			if errors.As(err, &mismatch) || errors.As(err, &refused) {
				// The server said no to the handshake itself. Retrying
				// cannot change its mind; the app needs updating (or the
				// server does).
				c.setStatus(gen, StatusVersionMismatch, err)
				return
			}
			if !c.setStatus(gen, StatusReconnecting, err) {
				return
			}
			if !c.wait(ctx, backoff.Next()) {
				return
			}
			continue
		}

		backoff.Reset()
		if !c.setStatus(gen, StatusConnected, nil) {
			_ = conn.Close()
			return
		}

		select {
		case <-ctx.Done():
			_ = conn.Close()
			return
		case <-conn.Done():
		}
		cause := conn.Err()
		c.mu.Lock()
		if gen == c.gen && c.conn == conn {
			c.conn = nil
		}
		c.mu.Unlock()
		if !c.setStatus(gen, StatusReconnecting, cause) {
			return
		}
		if !c.wait(ctx, backoff.Next()) {
			return
		}
	}
}

// wait sleeps out one backoff step, cut short by Retry or cancellation.
// Reports false when the loop should stop.
func (c *Connection) wait(ctx context.Context, d time.Duration) bool {
	timer := time.NewTimer(d)
	defer timer.Stop()
	select {
	case <-ctx.Done():
		return false
	case <-c.wake:
		return true
	case <-timer.C:
		return true
	}
}

// setStatus records a status change for generation gen, and reports false
// if that generation is no longer the current one (so the loop exits).
func (c *Connection) setStatus(gen int, status Status, err error) bool {
	c.mu.Lock()
	if gen != c.gen {
		c.mu.Unlock()
		return false
	}
	c.status = status
	c.lastErr = err
	c.mu.Unlock()
	c.notify()
	return true
}

// onMessage is the fold: runs on the socket's reader goroutine, one message
// at a time, under the lock a render pass also takes.
func (c *Connection) onMessage(gen int, msg any) {
	c.mu.Lock()
	if gen != c.gen {
		c.mu.Unlock()
		return
	}
	c.session.Apply(msg)
	if _, ok := msg.(*wire.Agents); ok {
		c.agentsAt = time.Now()
	}
	c.mu.Unlock()
	c.notify()
}

// Read runs fn with the session under the lock. Every render-pass read goes
// through here; fn must not block and must not call back into Connection.
func (c *Connection) Read(fn func(s *catsclient.Session)) {
	c.mu.Lock()
	defer c.mu.Unlock()
	fn(c.session)
}

// Info is a snapshot of the connection's state for a render pass.
type Info struct {
	Status   Status
	Err      error
	Endpoint catsclient.Endpoint
	CertSeen string
	// Caps is the server's advertised capability set, empty when not
	// connected. Screens gate window-following and pane-addressed input on
	// it.
	Caps []string
}

func (c *Connection) Info() Info {
	c.mu.Lock()
	defer c.mu.Unlock()
	info := Info{
		Status:   c.status,
		Err:      c.lastErr,
		Endpoint: c.endpoint,
		CertSeen: c.certSeen,
	}
	if c.conn != nil {
		info.Caps = c.conn.Caps()
	}
	return info
}

// HasCap reports whether the live server advertised a capability. False
// when disconnected.
func (c *Connection) HasCap(name string) bool {
	c.mu.Lock()
	defer c.mu.Unlock()
	return c.conn != nil && c.conn.HasCap(name)
}

// Live is the current Conn for issuing commands, or nil while disconnected.
// Callers hold it only for the duration of one command; a reconnect makes a
// new one.
func (c *Connection) Live() *catsclient.Conn {
	c.mu.Lock()
	defer c.mu.Unlock()
	if c.conn == nil || c.conn.IsClosed() {
		return nil
	}
	return c.conn
}

// AgentAge is how long an agent has been in its state, as of now: the
// rollup's since_ms plus the time since the rollup arrived. Negative
// since_ms means the age was never published; that is reported as -1 so a
// label can say nothing rather than "0s".
func (c *Connection) AgentAge(item wire.AgentItem) time.Duration {
	if item.SinceMs < 0 {
		return -1
	}
	c.mu.Lock()
	at := c.agentsAt
	c.mu.Unlock()
	age := time.Duration(item.SinceMs) * time.Millisecond
	if !at.IsZero() {
		age += time.Since(at)
	}
	return age
}

// isUnauthorized recognises a refused credential in a dial error.
//
// coder/websocket reports a non-101 upgrade as an error whose text carries
// the status code, and there is no typed error to unwrap for it. Matching on
// the code is a string test, and it is here in one place with this note so
// that a library change that reshapes the message breaks one function.
func isUnauthorized(err error) bool {
	if err == nil {
		return false
	}
	msg := err.Error()
	return strings.Contains(msg, "401") || strings.Contains(msg, "403")
}
